mocks/timestamp.js

```diff
@@ -19,34 +19,34 @@ class Timestamp {
   }
 
   toDate() {
-    const d = new Date(0);
-    d.setTime(this.seconds * 1000);
-    d.setMilliseconds(this.nanoseconds / 1000000);
-    return mockTimestampToDate(...arguments) || d;
+    return mockTimestampToDate(...arguments) || new Date(this._toMillis());
   }
 
   toMillis() {
-    const d = new Date(0);
-    d.setSeconds(this.seconds);
-    d.setMilliseconds(this.nanoseconds / 1000000);
-    return mockTimestampToMillis(...arguments) || d.getMilliseconds();
+    return mockTimestampToMillis(...arguments) || this._toMillis();
   }
 
   valueOf() {
     return JSON.stringify(this.toMillis());
   }
 
   static fromDate(date) {
-    return (
-      mockTimestampFromDate(...arguments) ||
-      new Timestamp(date.getTime() * 0.001, date.getMilliseconds() * 1000000)
-    );
+    return mockTimestampFromDate(...arguments) || Timestamp._fromMillis(date.getTime());
   }
 
   static fromMillis(millis) {
-    const d = new Date(0);
-    d.setMilliseconds(millis);
-    return mockTimestampFromMillis(...arguments) || Timestamp.fromDate(d);
+    return mockTimestampFromMillis(...arguments) || Timestamp._fromMillis(millis);
+  }
+
+  static _fromMillis(millis) {
+    const seconds = Math.floor(millis / 1000);
+    const nanoseconds = 1000000 * (millis % 1000);
```

Comment:
If `millis < 0`, then `nanoseconds` may also be a negative number. This should work for both negative and positive values of `millis`:
```suggestion
    const nanoseconds = 1000000 * (millis - seconds * 1000);
```

Reply:
Our test includes a negative number.  I think I better check that it works as expected.  I feel (vaguely) that this is something I ran across originally, so while your suggestion looks correct, I'll try putting in some additional tests (next break in work) to verify it.

Reply:
Yes, but only the seconds part in the test data is nonzero, so this line doesn't come into play.

Reply:
Yup, your right! Change in there along with a small test.